handlers/secrets.go

```diff
@@ -71,33 +71,65 @@ func UpdateSecrets(request requests.CreateFunctionRequest, deployment *v1beta1.D
 		}
 	}
 
-	if len(secretVolumeProjections) > 0 {
-		volumeName := fmt.Sprintf("%s-projected-secrets", request.Service)
-		projectedSecrets := apiv1.Volume{
-			Name: volumeName,
-			VolumeSource: apiv1.VolumeSource{
-				Projected: &apiv1.ProjectedVolumeSource{
-					Sources: secretVolumeProjections,
-				},
+	volumeName := fmt.Sprintf("%s-projected-secrets", request.Service)
+	projectedSecrets := apiv1.Volume{
+		Name: volumeName,
+		VolumeSource: apiv1.VolumeSource{
+			Projected: &apiv1.ProjectedVolumeSource{
+				Sources: secretVolumeProjections,
 			},
-		}
+		},
+	}
+
+	// remove the existing secrets volume, if we can find it. The update volume will be
+	// added below
+	existingVolumes := removeVolume(volumeName, deployment.Spec.Template.Spec.Volumes)
+	deployment.Spec.Template.Spec.Volumes = existingVolumes
+	if len(secretVolumeProjections) > 0 {
+		deployment.Spec.Template.Spec.Volumes = append(existingVolumes, projectedSecrets)
+	}
 
-		deployment.Spec.Template.Spec.Volumes = append(deployment.Spec.Template.Spec.Volumes, projectedSecrets)
+	// add mount secret as a file
+	updatedContainers := []apiv1.Container{}
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		mount := apiv1.VolumeMount{
+			Name:      volumeName,
+			ReadOnly:  true,
+			MountPath: "/run/secrets",
+		}
 
-		// add mount secret as a file
-		updatedContainers := []apiv1.Container{}
-		for _, container := range deployment.Spec.Template.Spec.Containers {
-			mount := apiv1.VolumeMount{
-				Name:      volumeName,
-				ReadOnly:  true,
-				MountPath: "/run/secrets",
-			}
+		// remove the existing secrets volume mount, if we can find it. We update it later.
+		container.VolumeMounts = removeVolumeMount(volumeName, container.VolumeMounts)
+		if len(secretVolumeProjections) > 0 {
 			container.VolumeMounts = append(container.VolumeMounts, mount)
-			updatedContainers = append(updatedContainers, container)
 		}
 
-		deployment.Spec.Template.Spec.Containers = updatedContainers
+		updatedContainers = append(updatedContainers, container)
 	}
 
+	deployment.Spec.Template.Spec.Containers = updatedContainers
+
 	return nil
 }
+
+// removeVolume returns a Volume slice with any volumes matching volumeName removed
+func removeVolume(volumeName string, volumes []apiv1.Volume) []apiv1.Volume {
+	for i, v := range volumes {
+		if v.Name == volumeName {
+			volumes = append(volumes[:i], volumes[i+1:]...)
```

Comment:
@johnmccabe left a comment about running a break at the point where you've matched?

Reply:
Yes, I responded here https://github.com/openfaas/faas-netes/pull/192/files/b482effe30476916c17a20c17b436cab9e8600a3#r186285541

We can break the loop, but that will not ensure that there is exactly one instance of that volume in the list, it will only remove the first one found.  I think visiting the entire slice is the most accurate.

Reply:
Yeah this sounds ok.

Reply:
I added a new test to cover this edge case.  Adding the test revealed a bug in the filter implementation.  It will cause an index error because we are changing the length of the slice while iterating over it.  The new method is simpler and more standard filter method.  All tests are passing now.